tools/autograd/templates/Functions.cpp

```diff
@@ -1800,44 +1800,65 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
 // Invertible case is derived from Jacobi's formula, and also can be found at:
 // http://eprints.maths.ox.ac.uk/1079/1/NA-08-01.pdf
 Tensor det_backward(const Tensor & grad, const Tensor& self, const Tensor& det) {
-  auto det_val = det.item<double>();
-  if (det_val != 0 /* invertible */) {
-    return grad * det * self.inverse().t();
-  } else /* otherwise det = \prod(sigma) = 0, use svd */ {
+  auto singular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
     Tensor u, sigma, v;
     std::tie(u, sigma, v) = self.svd();
-    auto gsigma = prod_backward(grad, sigma, det);
+    auto gsigma = prod_backward(grad.unsqueeze(-1), sigma, det.unsqueeze(-1));
     return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+  };
+
+  if ((det != 0).all().item<uint8_t>()) {  // all matrices are invertible
+    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  } else if ((det == 0).all().item<uint8_t>()) {  // all matrices are singular
+    return singular_case_backward(grad, self, det);
+  } else {  // some matrices are invertible, some matrices are singular
+    return at::where(det != 0,
```

Comment:
I'm still not a fan of this. This means that just running det in a loop and cat afterwards can often be much faster.

You are synching by accessing det.all() anyways. So, why not just do a `det.nonzero()`, selecting and calculating both parts, and then putting them together?

Reply:
I actually didn’t think of nonzero, I’ll try it out.

Reply:
yeah, thanks! Inplace `index_put` gradients into correct locations may break double backward, you might also want to check if that still works.

Reply:
Inplace index_put has a gradient defined, so I don't think it'll break double backward. I've made changes as per your suggestions.